### ISSUE TITLE
feature request: allow custom pagination component

### ISSUE THREAD

**Issue by dgreene1** [grommet/grommet#6205]
### Expected Behavior

The `paginate` prop (in addition to it's current true and false value) would allow a custom component to be passed. We'd have to discuss the properties that this custom component would be passed, but I expect it would be the [same exact props that your internal paginator accepts](https://github.com/grommet/grommet/blob/9cb1b5d6f10e8224479761464263f5f8ba13625e/src/js/components/Pagination/index.d.ts#L9).

The goal of this is so that we can style the paginator as our company would like to and also so we can add additional important features like allowing the user to change the size of the page.

The is how our UX department would like us to render the pagination.

![image](https://user-images.githubusercontent.com/14020024/175973904-d971ed5e-89b6-4a89-8be8-45c99ca5d892.png)

So a compromise might be to allow "slots" (sorry to use an old Angular term) that would allow us to put something to the left and to the right of pagination. But I think allowing a custom component would be the most flexible solution. 

In summary, this is what I imagine the final type would look like:

```ts
export type paginatePropType = boolean | PaginationProps | (props: PaginationProps) => React.Element
```

### Actual Behavior

The paginator widget can not be swapped.

### URL, screen shot, or Codepen exhibiting the issue

see above

### Steps to Reproduce

n.a.

### Your Environment

n.a.

**Comment by halocline** [comment]
Does consuming `<Pagination />` directly as shown in this [Codesandbox](https://codesandbox.io/s/pagination-prop-and-style-ubf667?file=/src/App.js) meet your needs?

I think the only aspect not currently customizable from your design is the separator style which would be a minor enhancement.

**Comment by dgreene1** [comment]
I think that approach would work if we (our wrapper library) ensures that the component we expose doesn't allow the user to turn on pagination and the `onMore` prop since the two would not work together well at all.

So the ideal solution would be the proposal above since then there would be automated tests inside of Grommet's codebase that would be enshrine the right to use custom pagination. My fear of using your approach described in https://github.com/grommet/grommet/issues/6205#issuecomment-1167899148 is that there could be a change in Grommet that causes a regression in the consumer code but that wouldn't be noticed in any Grommet PR. Have I communicated my point clearly?

**Comment by jcfilben** [comment]
I think the `paginate` prop on DataTable could be enhanced to accept a function, I think this would allow for the flexibility needed. We follow a similar pattern on Calendar with the `header` prop where `header` can be a function with the following passed to it: `date, locale, onPreviousMonth, onNextMonth, previousInBound, nextInBound`. For DataTable `paginate` we would need to pass something like `data, page, step` (maybe more) to the function. @dgreene1 would an enhancement like this work for your use case?

**Comment by dgreene1** [comment]
@jcfilben the enhancement as you describe it would work beautifully.

The only thing I’m not sure of is how we would update the pagination when the filter or sort state changes the count or the current page that item would be on. AntD has an on change event that allows the user to hook into this but I don’t know how necessary that is with Grommet due to my current lack of experience on how pages are managed in state with filters.

cc’ing @mellis481 who also works on my team and might be able to provide more guidance.

But yes this would work great @jcfilben. And I’m happy to review the PR to observe the filter stuff I’m talking about and maybe even add a unit test once I’m back from vacation.
